pgml-apps/pgml-chat/pgml_chat/main.py

```diff
@@ -117,13 +234,112 @@ async def upsert_documents(folder: str) -> int:
+    temperature=0.7,
+    max_tokens=256,
+    top_p=0.9,
+    user_name="",
+):
+    messages = []
+    messages.append({"role": "system", "content": system_prompt})
+
+    chat_history_messages = await chat_collection.get_documents( {
+        "limit" : chat_history*2,
+        "order_by": {"timestamp": "desc"},
+        "filter": {
+            "metadata": {
+                "$and" : [
+                    {
+                        "$or": 
+                        [
+                        {"role": {"$eq": "assistant"}},
+                        {"role": {"$eq": "user"}}
+                        ]
+                    },
+                    {
+                        "interface" : {
+                            "$eq" : chat_interface
+                        }
+                    },
+                    {
+                        "user_name" : {
+                            "$eq" : user_name
+                        }
+                    }
+                ]
+            }
+        }
+        }
+    )
+    
+    # Reverse the order so that user messages are first
+
+    chat_history_messages.reverse()
+
+    conversation_history = ""
+    for entry in chat_history_messages:
+        document = entry["document"]
+        if document["role"] == "user":
+            conversation_history += "User: " + document["text"] + "\n"
+        if document["role"] == "assistant":
+            conversation_history += "Assistant: " + document["text"] + "\n"
+
+    log.info(conversation_history)
+
+    history_documents = []
+    user_message_id = str(uuid4())[:8]
+    _document = {
+        "text": user_input,
+        "id": user_message_id,
+        "interface": chat_interface,
+        "role": "user",
+        "timestamp": pendulum.now().timestamp(),
+        "user_name": user_name,
+    }
+    history_documents.append(_document)
+
+    if user_input:
+        query = await get_prompt(user_input,conversation_history)
+
+    messages.append({"role": "user", "content": query})
+
+    log.info(messages)
+
+    response = await generate_response(
+        messages,
+        openai_api_key,
+        max_tokens=max_tokens,
+        temperature=temperature,
+        top_p=top_p,
+    )
+
+    _document = {
+        "text": response,
+        "id": str(uuid4())[:8],
+        "parent_message_id" : user_message_id,
+        "interface": chat_interface,
+        "role": "assistant",
+        "timestamp": pendulum.now().timestamp(),
+        "user_name": user_name,
+    }
+    history_documents.append(_document)
+
+    await chat_collection.upsert_documents(history_documents)
+
+    return response
+
+
 async def generate_response(
     messages, openai_api_key, temperature=0.7, max_tokens=256, top_p=0.9
 ):
     openai.api_key = openai_api_key
     log.debug("Generating response from OpenAI API: " + str(messages))
     response = openai.ChatCompletion.create(
-        model="gpt-3.5-turbo",
+        # model="gpt-3.5-turbo-16k",
+        model="gpt-4",
```

Comment:
Who's paying for this?

Reply:
OPENAI_API_KEY is an environment variable. User pays for it.